slictl/slictl.8
.\" $Id$
.\" %GPL_START_LICENSE%
.\" ---------------------------------------------------------------------
.\" Copyright 2015-2016, Google, Inc.
.\" Copyright 2008-2018, Pittsburgh Supercomputing Center
.\" All rights reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or (at
.\" your option) any later version.
.\"
.\" This program is distributed WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE.  See the GNU General Public License contained in the file
.\" `COPYING-GPL' at the top of this distribution or at
.\" https://www.gnu.org/licenses/gpl-2.0.html for more details.
.\" ---------------------------------------------------------------------
.\" %END_LICENSE%
.\" %PFL_MODULES ctl rpc %
.Dd April 19, 2016
.Dt SLICTL 8
.ds volume PSC \- SLASH2 Administrator's Manual
.Os http://www.psc.edu/
.Sh NAME
.Nm slictl
.Nd
.Tn SLASH2
.Xr sliod 8
runtime control
.Sh SYNOPSIS
.Nm slictl
.Op Fl HInV
.Op Fl p Ar paramspec
.Op Fl S Ar socket
.Op Fl s Ar showspec
.Op Ar cmd arg ...
.Sh DESCRIPTION
The
.Nm
utility examines and manipulates
.Xr sliod 8
behavior.
.Pp
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/cmd.mdoc {
.\"	cmds => {
.\"		"export Oo Fl Rv\\& Oc Ar src ... dst" => <<'EOF',
.\"			Copy the specified files from the
.\"			.Tn SLASH2
.\"			file system to the destination specified at
.\"			.Ar dst
.\"			on this machine's local file system.
.\"			The
.\"			.Ar src
.\"			arguments are relative to the
.\"			.Tn SLASH2
.\"			name space and must have residency on this
.\"			.Tn I/O
.\"			node's backing file system.
.\"			.Pp
.\"			The options are as follows:
.\"			.Bl -tag -offset 3n -width 3n
.\"			.It Fl R
.\"			Recursively traverse all file hierarchies rooted at the specified
.\"			.Ar src
.\"			nodes.
.\"			.It Fl v
.\"			Display verbose information about processing.
.\"			.El
.\"			EOF
.\"		"import Oo Fl Rvx\\& Oc Oo Fl F Ar pfid Oc Ar file ... dst" => <<'EOF',
.\"			Import the specified files into the
.\"			.Tn SLASH2
.\"			file system name space rooted at
.\"			.Ar dst .
.\"			Importing publishes a read-only copy of the data
.\"			to the local metadata server.
.\"			Reimporting an existing file refreshes its file
.\"			attributes on the metadata server.
.\"			.Pp
.\"			The options are as follows:
.\"			.Bl -tag -offset 3n -width 3n
.\"			.It Fl F Ar pfid
.\"			If specified,
.\"			.Ar dst
.\"			is used as the basename to import the given source file into the
.\"			directory in the
.\"			.Tn SLASH2
.\"			namespace with the given
.\"			.Tn FID
.\"			.Ar pfid .
.\"			.It Fl R
.\"			Recursively traverse all file hierarchies rooted at the specified
.\"			.Ar file
.\"			nodes.
.\"			.It Fl v
.\"			Display verbose information about processing.
.\"			.It Fl x
.\"			When
.\"			.Ar dst
.\"			already exists, instead of refreshing
.\"			attributes, register a residency state
.\"			for this
.\"			.Tn I/O
.\"			node in the replica table for the specified files.
.\"			.El
.\"			.Pp
.\"			Note: usage for directories is not similar to
.\"			.Xr cp 1
.\"			due to ambiguities regarding
.\"			.Ar dst
.\"			existence;
.\"			rather, intended destination paths must be
.\"			supplied in full.
.\"			EOF
.\"		stop => <<'EOF',
.\"			Instruct
.\"			.Xr sliod 8
.\"			to completely terminate daemon execution.
.\"			EOF
.\"	}
The supported commands are as follows:
.Bl -tag -width 3n
.It Cm export Oo Fl Rv\& Oc Ar src ... dst
Copy the specified files from the
.Tn SLASH2
file system to the destination specified at
.Ar dst
on this machine's local file system.
The
.Ar src
arguments are relative to the
.Tn SLASH2
name space and must have residency on this
.Tn I/O
node's backing file system.
.Pp
The options are as follows:
.Bl -tag -offset 3n -width 3n
.It Fl R
Recursively traverse all file hierarchies rooted at the specified
.Ar src
nodes.
.It Fl v
Display verbose information about processing.
.El
.It Cm import Oo Fl Rvx\& Oc Oo Fl F Ar pfid Oc Ar file ... dst
Import the specified files into the
.Tn SLASH2
file system name space rooted at
.Ar dst .
Importing publishes a read-only copy of the data
to the local metadata server.
Reimporting an existing file refreshes its file
attributes on the metadata server.
.Pp
The options are as follows:
.Bl -tag -offset 3n -width 3n
.It Fl F Ar pfid
If specified,
.Ar dst
is used as the basename to import the given source file into the
directory in the
.Tn SLASH2
namespace with the given
.Tn FID
.Ar pfid .
.It Fl R
Recursively traverse all file hierarchies rooted at the specified
.Ar file
nodes.
.It Fl v
Display verbose information about processing.
.It Fl x
When
.Ar dst
already exists, instead of refreshing
attributes, register a residency state
for this
.Tn I/O
node in the replica table for the specified files.
.El
.Pp
Note: usage for directories is not similar to
.Xr cp 1
due to ambiguities regarding
.Ar dst
existence;
rather, intended destination paths must be
supplied in full.
.It Cm stop
Instruct
.Xr sliod 8
to completely terminate daemon execution.
.El
.\" }%
.Pp
The options are as follows:
.Bl -tag -width 3n
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/H.mdoc {
.It Fl H
Suppress headers when showing data displays.
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/I.mdoc {
.It Fl I
Display large numerical values accurately instead of rounded with
postfixes of demonination such as
.Sq G
for gibibytes,
.Sq M
for mebibytes, etc.\&
.Pq Do inhuman Dc number mode .
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/n.mdoc {
.It Fl n
Suppress
.Tn DNS
lookups for network addresses.
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/p.mdoc {
.\"	log_xr => "in\n.Xr sliod 8\n",
.\"	params => {
.\"		'pid' => "Daemon system process ID.",
.\"		'sys.bminseqno'
.\"		     => "Bmap lease minimum sequence number to allow.",
.\"		'sys.reclaim_batchno'
.\"		     => "Highest observed garbage reclamation batch number.",
.\"		'sys.reclaim_xid'
.\"		     => "Highest observed garbage reclamation batch transaction ID.",
.\"		'sys.sync_max_writes'
.\"		     => "Number of incoming writes to receive on a file from\n" .
.\"			"clients before the data synchronizer begins\n" .
.\"			"flushing to backing store.",
.\"		'sys.selftestrc'
.\"		     => "Error status of last backend file system health check.",
.\"		'sys.nbrq_outstanding'
.\"		     => "Number of currently outstanding asynchronous RPCs.",
.\"	},
.It Fl p Ar paramspec
Query/manipulate operational parameters.
.Ar paramspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Sm off
.Oo Ar thread-name Ns Li .\& Oc Ar param
.Op Oo Li +- Oc Li = Ar value
.Sm on
.Ed
.Pp
Some parameters may be read- or write-only.
Some support modification by the assignment operators
.Li +=
and
.Li -= .
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread-name .
The default behavior is to apply the operation to all applicable threads.
.Pp
.Ar param
may be one of the following:
.Bl -tag -width 1n -offset 3n
.It Cm lnet.networks
.Tn LNET
network configuration.
.It Cm lnet.port
If applicable,
.Xr tcp 7
port to use to connect to remote
.Tn LNET
peers.
.It Cm log.file
File path name where log messages will be written.
This field is write-only.
If the
.Li +=
assignment operator is used, this file will not be truncated.
.It Cm log.format
The header prepended to server log messages.
See the description of
.Ev PSC_LOG_FORMAT
in
.Xr sliod 8
for details on this format.
.It Cm log.level Ns Op . Ns Ar subsystem
The logging level of debug message output.
.Pp
See the description of
.Ev PSC_LOG_LEVEL_ Ns Ar subsys
in
.Xr sliod 8
for a list of available subsystems.
If
.Ar subsystem
is not specified, all subsystems will be accessed.
.Pp
See the description of
.Ev PSC_LOG_LEVEL
in
.Xr sliod 8
for recognized log level values.
.It Cm log.points
List of places in the source code (in the form
.Dq file:line-number )
where granular logging can be enabled on a case-by-case
basis.
Usually assigned to by the
.Li +=
operator.
.It Cm pid
Daemon system process ID.
.It Cm pool. Ns Ar name
Access the memory pool specified by
.Ar name .
The following sub-fields are available:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm free
Current number of allocated entries available for use in pool.
.It Cm max
Upper bound for number of entries to which auto-sized pools can grow.
.It Cm min
Lower bound for number of entries to which auto-sized pools can shrink.
.It Cm reap
Assign a value to reap the specified number of entries.
Only applies to pools that have a reaper routine.
This field is write-only.
.It Cm thres
Threshold for unused items for auto-sized pools before items are freed.
.It Cm total
Current number of entries contained in pool.
.El
.It Cm rlim
Process resource limits.
See
.Xr getrlimit 2
or
.Xr ulimit 1
for more information.
.Pp
.Bl -column "msgqueue" -offset 3n
.It Sy Name Ta Sy Field
.It =============================
.It Cm cpu        Ta Dv RLIMIT_CPU
.It Cm csize      Ta Dv RLIMIT_CORE
.It Cm dsize      Ta Dv RLIMIT_DATA
.It Cm fsize      Ta Dv RLIMIT_FSIZE
.It Cm locks      Ta Dv RLIMIT_LOCKS
.It Cm maxproc    Ta Dv RLIMIT_NPROC
.It Cm mem        Ta Dv RLIMIT_AS
.It Cm mlock      Ta Dv RLIMIT_MEMLOCK
.It Cm msgqueue   Ta Dv RLIMIT_MSGQUEUE
.It Cm nice       Ta Dv RLIMIT_NICE
.It Cm nofile     Ta Dv RLIMIT_NOFILE
.It Cm rss        Ta Dv RLIMIT_RSS
.It Cm rtprio     Ta Dv RLIMIT_RTPRIO
.It Cm rttime     Ta Dv RLIMIT_RTTIME
.It Cm sigpndg    Ta Dv RLIMIT_SIGPENDING
.It Cm stksize    Ta Dv RLIMIT_STACK
.El
.It Cm rusage
Process resource usage information.
See
.Xr getrusage 2 .
.It Cm sys.bminseqno
Bmap lease minimum sequence number to allow.
.It Cm sys.nbrq_outstanding
Number of currently outstanding asynchronous RPCs.
.It Cm sys.reclaim_batchno
Highest observed garbage reclamation batch number.
.It Cm sys.reclaim_xid
Highest observed garbage reclamation batch transaction ID.
.It Cm sys.selftestrc
Error status of last backend file system health check.
.It Cm sys.sync_max_writes
Number of incoming writes to receive on a file from
clients before the data synchronizer begins
flushing to backing store.
.El
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/S.mdoc {
.\"	daemon	=> qq{sliod},
.\"	sock	=> "/var/run/sliod. Ns Ic %h Ns Pa .sock"
.It Fl S Ar socket
Specify an alternative socket file.
The following tokens are replaced in the file name specified:
.Pp
.Bl -tag -offset 3n -width Ds -compact
.It Cm %h
the machine hostname
.It Cm %n
the daemon executable base name, i.e.\&
.Dq sliod
.It Cm %%
a literal
.Sq %
character
.El
.Pp
The default is
.Pa /var/run/sliod. Ns Ic %h Ns Pa .sock .
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/show.mdoc {
.\"	show => {
.\"		bmap		=> "In-memory bmaps",
.\"		connections	=> "Status of\n.Tn SLASH2\npeers on network",
.\"		fidcache	=> ".Tn FID\n.Pq file- Ns Tn ID\ncache members",
.\"		replwkst	=> "Status of active replications",
.\"		slvrs		=> "In-memory slivers (bmap slices)",
.\"	},
.\"	pools => {
.\"		bmap		=> "Block map structures",
.\"	},
.\"	listcaches => {
.\"		bcr_ready	=> "Bmap CRC update tranmission queue",
.\"		breleaseq	=> "Bmaps awaiting release by MDS response",
.\"		crcqslvrs	=> "Bmap slivers awaiting checksumming",
.\"		fcmhidle	=> "Recently used files",
.\"		lruslvrs	=> "Recently used bmap slivers",
.\"		readaheadq	=> "Readahead I/O work queue",
.\"		replwkpnd	=> "Pending replication work",
.\"	},
.\"	hashtables => {
.\"		fidc		=> "files\n.Po file\n.Tn ID\ncache\n.Pc",
.\"		res		=> "network resources\n.Pq network Tn ID",
.\"		rpcconn		=> "network resources\n.Pq network Tn ID",
.\"	}
.It Fl s Ar showspec
Show values.
.Ar showspec
has the following format:
.Bd -unfilled -offset 3n
.Sm off
.Ar param
.Op : Ar subspec
.Sm on
.Ed
.Pp
.Ar param
may be specified as any non-ambiguous prefix abbreviation of the
following:
.Pp
.Bl -tag -width 1n -offset 3n
.It Cm bmap
In-memory bmaps
.It Cm connections
Status of
.Tn SLASH2
peers on network
.It Cm fidcache
.Tn FID
.Pq file- Ns Tn ID
cache members
.It Cm hashtables
Hash table statistics.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar hash-table Ns Op , Ns Ar ...
.Ed
.Pp
.Ar hash-table
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm fidc
files
.Po file
.Tn ID
cache
.Pc
.It Cm res
network resources
.Pq network Tn ID
.It Cm rpcconn
network resources
.Pq network Tn ID
.El
.Pp
If
.Ar subspec
is left unspecified, all hash tables will be accessed.
.It Cm listcaches
List cache statistics.
.Ar subspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Ar list Ns Op , Ns Ar ...
.Ed
.Pp
.Ar list
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm bcr_ready
Bmap CRC update tranmission queue
.It Cm breleaseq
Bmaps awaiting release by MDS response
.It Cm crcqslvrs
Bmap slivers awaiting checksumming
.It Cm fcmhidle
Recently used files
.It Cm lruslvrs
Recently used bmap slivers
.It Cm readaheadq
Readahead I/O work queue
.It Cm replwkpnd
Pending replication work
.El
.Pp
If
.Ar subspec
is left unspecified, all list caches will be accessed.
.It Cm lnetif
Lustre network interfaces.
.It Cm loglevels
Thread logging levels.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar thread Ns Op , Ns Ar ...
.Ed
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread .
If
.Ar subspec
is left unspecified, all threads will be accessed.
.It Cm opstats
Operation counters/statistics.
.Ar subspec
has the following format:
.Pp
.Bd -unfilled -offset 3n
.Ar opstats Ns Op , Ns Ar ...
.Ed
.Pp
.Ar opstats
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 3n
.It Cm lni-rcv- Ns Ar if ,
.It Cm lni-snd- Ns Ar if
Data sent/received per
.Tn LNET
networking interface.
.Pp
.It Cm lusklnd- Ns Ar mode Ns Cm -rcv ,
.It Cm lusklnd- Ns Ar mode Ns Cm -snd
Data sent/received over userland socket networking device.
.Ar mode
may be
.Cm pasv
.Pq passive
or
.Cm aggr
.Pq aggregate .
.Pp
.It Cm rpc- Ns Ar addr Ns Cm -rcv ,
.It Cm rpc- Ns Ar addr Ns Cm -snd
Data sent/received per
.Tn RPC
peer.
.Pp
.El
.Pp
If
.Ar subspec
is left unspecified, all
.Tn I/O
statistics will be accessed.
.It Cm pools
Memory pool statistics.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar pool Ns Op , Ns Ar ...
.Ed
.Pp
.Ar pool
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm bmap
Block map structures
.El
.Pp
If
.Ar subspec
is left unspecified, all pools will be accessed.
.It Cm replwkst
Status of active replications
.It Cm rpcrqs
Remote procedure calls (RPC).
.It Cm rpcsvcs
.Tn RPC
services.
.It Cm slvrs
In-memory slivers (bmap slices)
.It Cm threads
Daemon thread activity and statistics.
.Ar subspec
has the following format:
.Bd -unfilled -offset 3n
.Ar thread Ns Op , Ns Ar ...
.Ed
.Pp
See
.Sx Thread Specification
for details on specifying
.Ar thread .
If
.Ar subspec
is left unspecified, all threads will be accessed.
.It Cm workrqs
Work requests.
.El
.Pp
The special value
.Sq \&?
may also be specified to display a list of recognized values.
.\" }%
.It Fl V
Show version.
.El
.\" %PFL_INCLUDE $PFL_BASE/doc/pflctl/thr.mdoc {
.\"	thrs => {
.\"		"sliaiothr"			=> "Asynchronous\n.Tn I/O\nprocessor",
.\"		"slibmaprlsthr"			=> "Bmap releaser",
.\"		"slibreapthr"			=> "Bmap reaper",
.\"		"sliconnthr"			=> "Peer resource connection monitor",
.\"		"slicrudthr"			=> "Bmap CRC updater",
.\"		"slictlacthr"			=> ".Nm\nconnection acceptor",
.\"		"slictlthr"			=> ".Nm\nconnection processor",
.\"		"slifreapthr"			=> "File reaper",
.\"		"slihealththr"			=> "Backing file system health checker",
.\"		"slilnacthr- Ns Ar %s"		=> "Lustre network acceptor thread",
.\"		"slinbrqthr"			=> "Non-blocking\n.Tn RPC\nreply handler",
.\"		"sliopstimerthr"		=> qq{Internal operation count updater},
.\"		"slireplpndthr"			=> "Pending replication work processor",
.\"		"sliricthr Ns Ar %02d"		=> "Client\n.Tn RPC\nrequest service thread",
.\"		"sliriithr Ns Ar %02d"		=> ".No Inter- Ns Tn I/O RPC\nrequest service thread",
.\"		"slirimthr Ns Ar %02d"		=> ".Tn MDS RPC\nrequest service thread",
.\"		"slislvrthr Ns Ar %d"		=> "Sliver monitoring thread",
.\"		"slistatfsthr"			=> "Periodic\n.Xr statvfs 2\nupdater",
.\"		"sliusklndplthr Ns Ar %d"	=> "Lustre userland socket poll thread",
.\"		"sliwkthr Ns Ar %d"		=> "Generic worker thread"
.\"	}
.Ss Thread Specification
Options which take
.Ar thread-name
parameters may be specified by one or more of the following tokens,
separated by commas:
.Pp
.Bl -tag -compact -offset 3n -width 16n
.It Cm sliaiothr
Asynchronous
.Tn I/O
processor
.It Cm slibmaprlsthr
Bmap releaser
.It Cm slibreapthr
Bmap reaper
.It Cm sliconnthr
Peer resource connection monitor
.It Cm slicrudthr
Bmap CRC updater
.It Cm slictlacthr
.Nm
connection acceptor
.It Cm slictlthr
.Nm
connection processor
.It Cm slifreapthr
File reaper
.It Cm slihealththr
Backing file system health checker
.It Cm slilnacthr- Ns Ar %s
Lustre network acceptor thread
.It Cm slinbrqthr
Non-blocking
.Tn RPC
reply handler
.It Cm sliopstimerthr
Internal operation count updater
.It Cm slireplpndthr
Pending replication work processor
.It Cm sliricthr Ns Ar %02d
Client
.Tn RPC
request service thread
.It Cm sliriithr Ns Ar %02d
.No Inter- Ns Tn I/O RPC
request service thread
.It Cm slirimthr Ns Ar %02d
.Tn MDS RPC
request service thread
.It Cm slislvrthr Ns Ar %d
Sliver monitoring thread
.It Cm slistatfsthr
Periodic
.Xr statvfs 2
updater
.It Cm sliusklndplthr Ns Ar %d
Lustre userland socket poll thread
.It Cm sliwkthr Ns Ar %d
Generic worker thread
.It Cm everyone
All threads
.Pq default, where applicable
.El
.\" }%
.\" %PFL_INCLUDE $PFL_BASE/doc/env.mdoc {
.Sh ENVIRONMENT
.Bl -tag -width 3n
.It Ev CTL_SOCK_FILE
Override the default control socket file path.
.El
.\" }%
.Sh FILES
.Bl -tag -width Pa
.It Pa /var/run/sliod. Ns Ic %h Ns Pa .sock
default
.Xr sliod 8
control socket
.El
.Sh SEE ALSO
.Xr sladm 7 ,
.Xr sliod 8


slictl/slictl.c
/* $Id$ */
/*
 * %GPL_START_LICENSE%
 * ---------------------------------------------------------------------
 * Copyright 2007-2018, Pittsburgh Supercomputing Center
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License contained in the file
 * `COPYING-GPL' at the top of this distribution or at
 * https://www.gnu.org/licenses/gpl-2.0.html for more details.
 * ---------------------------------------------------------------------
 * %END_LICENSE%
 */

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "pfl/cdefs.h"
#include "pfl/ctl.h"
#include "pfl/ctlcli.h"
#include "pfl/fmt.h"
#include "pfl/getopt.h"
#include "pfl/log.h"
#include "pfl/pfl.h"
#include "pfl/str.h"

#include "sliod/ctl_iod.h"
#include "sliod/slvr.h"

#include "ctl.h"
#include "slerr.h"
#include "ctlcli.h"
#include "pathnames.h"

const char	*daemon_name = "sliod";

void
packshow_bmaps(__unusedx char *bmaps)
{
	psc_ctlmsg_push(SLICMT_GETBMAP, sizeof(struct slctlmsg_bmap));
}

void
packshow_conns(__unusedx char *conn)
{
	psc_ctlmsg_push(SLICMT_GETCONN, sizeof(struct slctlmsg_conn));
}

void
packshow_replwkst(__unusedx char *fid)
{
	psc_ctlmsg_push(SLICMT_GET_REPLWKST, sizeof(struct slictlmsg_replwkst));
}

void
packshow_fcmhs(__unusedx char *fid)
{
	struct slctlmsg_fcmh *scf;

	scf = psc_ctlmsg_push(SLICMT_GETFCMH, sizeof(*scf));
	scf->scf_fg.fg_fid = FID_ANY;
}

int
replwkst_prhdr(__unusedx struct psc_ctlmsghdr *mh, __unusedx const void *m)
{
	printf("%-16s %6s %28s %4s %7s %7s %6s\n",
	    "replwk-stat-fid", "bmap#", "peer", "refs", "xfer", "total", "%prog");
	return(PSC_CTL_DISPLAY_WIDTH);
}

void
replwkst_prdat(__unusedx const struct psc_ctlmsghdr *mh, const void *m)
{
	char totbuf[PSCFMT_HUMAN_BUFSIZ], curbuf[PSCFMT_HUMAN_BUFSIZ];
	char rbuf[PSCFMT_RATIO_BUFSIZ];
	const struct slictlmsg_replwkst *srws = m;

	pfl_fmt_ratio(rbuf, srws->srws_data_cur, srws->srws_data_tot);
	printf("%016"SLPRIxFID" %6d %28s %4d ",
	    srws->srws_fg.fg_fid, srws->srws_bmapno,
	    srws->srws_peer_addr, srws->srws_refcnt);
	if (psc_ctl_inhuman)
		printf("%7d %7d", srws->srws_data_cur,
		    srws->srws_data_tot);
	else {
		pfl_fmt_human(totbuf, srws->srws_data_tot);
		pfl_fmt_human(curbuf, srws->srws_data_cur);
		printf("%7s %7s", curbuf, totbuf);
	}
	printf(" %6s\n", rbuf);
}

int
slvr_prhdr(__unusedx struct psc_ctlmsghdr *mh, __unusedx const void *m)
{
	printf("%-16s %6s %3s %4s %7s %5s %9s\n",
	    "slvr-fid", "bmap#", "sl#", "refs", "flags", "err", "time");
	return(PSC_CTL_DISPLAY_WIDTH);
}

void
slvr_prdat(__unusedx const struct psc_ctlmsghdr *mh, const void *m)
{
	const struct slictlmsg_slvr *ss = m;

	printf("%016"SLPRIxFID" %6d %3d %4d %c%c%c%c%c%c "
	    "%5d %9"PRId64" \n",
	    ss->ss_fid, ss->ss_bno, ss->ss_slvrno, ss->ss_refcnt,
	    ss->ss_flags & SLVRF_FAULTING	? 'f' : '-',
	    ss->ss_flags & SLVRF_DATARDY	? 'd' : '-',
	    ss->ss_flags & SLVRF_DATAERR	? 'E' : '-',
	    ss->ss_flags & SLVRF_LRU		? 'l' : '-',
	    ss->ss_flags & SLVRF_FREEING	? 'F' : '-',
	    ss->ss_flags & SLVRF_ACCESSED	? 'a' : '-',
	    ss->ss_err, ss->ss_ts.tv_sec);
}

void
slictlcmd_export(int ac, char *av[])
{
	int i, c, recursive = 0, verbose = 0;
	struct slictlmsg_fileop *sfop;

	PFL_OPT_RESET();
	while ((c = getopt(ac, av, "+Rv")) != -1)
		switch (c) {
		case 'R':
			recursive = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			goto usage;
		}
	ac -= optind;
	av += optind;

	if (ac < 2)
 usage:
		errx(1, "usage: export [-Rv] file ... dst");
	for (i = 0; i < ac - 1; i++) {
		sfop = psc_ctlmsg_push(SLICMT_EXPORT, sizeof(*sfop));
		if (recursive)
			sfop->sfop_flags |= SLI_CTL_FOPF_RECURSIVE;
		if (verbose)
			sfop->sfop_flags |= SLI_CTL_FOPF_VERBOSE;
		strlcpy(sfop->sfop_fn, av[i], sizeof(sfop->sfop_fn));
		strlcpy(sfop->sfop_fn2, av[ac - 1], sizeof(sfop->sfop_fn));
	}
}

void
slictlcmd_import(int ac, char *av[])
{
	int i, c, recursive = 0, verbose = 0, xrepl = 0;
	struct slictlmsg_fileop *sfop;
	char fn[PATH_MAX], *endp;
	slfid_t pfid = FID_ANY;

	PFL_OPT_RESET();
	while ((c = getopt(ac, av, "+F:Rvx")) != -1)
		switch (c) {
		case 'F':
			pfid = strtoull(optarg, &endp, 0);
			if (endp == optarg || *endp)
				errx(1, "%s: invalid parent FID", optarg);
			break;
		case 'R':
			recursive = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'x':
			xrepl = 1;
			break;
		default:
			goto usage;
		}
	ac -= optind;
	av += optind;

	if (ac < 2)
 usage:
		errx(1, "usage: import [-Rvx] [-F pfid] file ... dst");
	for (i = 0; i < ac - 1; i++) {
		if (realpath(av[i], fn) == NULL) {
			warn("%s", av[i]);
			continue;
		}
		sfop = psc_ctlmsg_push(SLICMT_IMPORT, sizeof(*sfop));
		if (recursive)
			sfop->sfop_flags |= SLI_CTL_FOPF_RECURSIVE;
		if (verbose)
			sfop->sfop_flags |= SLI_CTL_FOPF_VERBOSE;
		if (xrepl)
			sfop->sfop_flags |= SLI_CTL_FOPF_XREPL;
		strlcpy(sfop->sfop_fn, fn, sizeof(sfop->sfop_fn));
		strlcpy(sfop->sfop_fn2, av[ac - 1],
		    sizeof(sfop->sfop_fn2));
		sfop->sfop_pfid = pfid;
	}
}

void
slictlcmd_stop(int ac, __unusedx char *av[])
{
	if (ac > 1)
		errx(1, "stop: invalid arguments");
	psc_ctlmsg_push(SLICMT_STOP, 0);
}

struct psc_ctlshow_ent psc_ctlshow_tab[] = {
	PSC_CTLSHOW_DEFS,
	{ "bmaps",		packshow_bmaps },
	{ "connections",	packshow_conns },
	{ "fcmhs",		packshow_fcmhs },
	{ "replwkst",		packshow_replwkst },

	/* aliases */
	{ "conns",		packshow_conns },
	{ "files",		packshow_fcmhs },
	{ "fidcache",		packshow_fcmhs }
};

struct psc_ctlmsg_prfmt psc_ctlmsg_prfmts[] = {
	PSC_CTLMSG_PRFMT_DEFS,
	{ replwkst_prhdr,	replwkst_prdat,		sizeof(struct slictlmsg_replwkst),	NULL },
	{ sl_conn_prhdr,	sl_conn_prdat,		sizeof(struct slctlmsg_conn),		NULL },
	{ sl_fcmh_prhdr,	sl_fcmh_prdat,		sizeof(struct slctlmsg_fcmh),		NULL },
	{ NULL,			NULL,			sizeof(struct slictlmsg_fileop),	NULL },
	{ NULL,			NULL,			sizeof(struct slictlmsg_fileop),	NULL },
	{ NULL,			NULL,			0,					NULL },
	{ sl_bmap_prhdr,	sl_bmap_prdat,		sizeof(struct slctlmsg_bmap),		NULL },
	{ slvr_prhdr,		slvr_prdat,		sizeof(struct slictlmsg_slvr),		NULL }
};

struct psc_ctlcmd_req psc_ctlcmd_reqs[] = {
	{ "export",	slictlcmd_export },
	{ "import",	slictlcmd_import },
	{ "stop",	slictlcmd_stop }
};

PFLCTL_CLI_DEFS;

__dead void
usage(void)
{
	extern const char *__progname;

	fprintf(stderr,
	    "usage: %s [-HInV] [-p paramspec] [-S socket] [-s value] [cmd arg ...]\n",
	    __progname);
	exit(1);
}

void
slictl_show_version(void)
{
	fprintf(stderr, "%d\n", sl_stk_version);
}

struct psc_ctlopt opts[] = {
	{ 'H', PCOF_FLAG, &psc_ctl_noheader },
	{ 'I', PCOF_FLAG, &psc_ctl_inhuman },
	{ 'n', PCOF_FLAG, &psc_ctl_nodns },
	{ 'p', PCOF_FUNC, psc_ctlparse_param },
	{ 's', PCOF_FUNC, psc_ctlparse_show },
	{ 'V', PCOF_FLAG, slictl_show_version },
};

int
main(int argc, char *argv[])
{
	pfl_init();
	sl_errno_init();
	pscthr_init(0, NULL, 0, "slictl");

	psc_ctlcli_main(SL_PATH_SLICTLSOCK, argc, argv, opts,
	    nitems(opts));
	exit(0);
}
